File: SupervisedLearning/FeedforwardNeuralNetwork/ai/neuron.go
package ai

import (
	"MyFeedforwardNeuralNetwork/utils"
	"math/rand"
	"time"
)

type Neuron struct {
	weights []float64 // the neuron's weight values
	bias    float64   // the neuron's bias value
}

func NewNeuron(numInputs int) *Neuron {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	weights := make([]float64, numInputs)
	for i := range weights {
		weights[i] = r.Float64()*2 - 1
	}

	bias := r.Float64()*0.01 - 0.005

	return &Neuron{weights, bias}
}

func (n *Neuron) Activate(inputs []float64) float64 {
	// Ensure the input length matches the number of weights
	if len(inputs) != len(n.weights) {
		panic("number of inputs does not match number of weights")
	}

	// Calculate the weighted sum of inputs and weights
	sum := n.bias
	for i := range inputs {
		sum += inputs[i] * n.weights[i]
	}

	return utils.Sigmoid(sum)
}

func (n *Neuron) UpdateWeights(inputs []float64, delta, learningRate float64) {
	for i := range n.weights {
		n.weights[i] += learningRate * delta * inputs[i]
	}

	n.bias += learningRate * delta
}

File: SupervisedLearning/classification/xor_ruby/ai/neuron.rb
class Neuron
  attr_accessor :weights

  def initialize(num_inputs)
    @weights = Array.new(num_inputs) { rand(-1.0..1.0) }
    @bias = rand(0.1..1.0) * 0.01
    @num_inputs = num_inputs
  end

  def activate(inputs)
    if @num_inputs != inputs.length then
      abort("Wrong number of inputs")
    end

    sum = @bias
    inputs.each_with_index do |input, i|
      sum += input * weights[i]
    end

    sigmoid(sum)
  end

  def calculate_error(errors, delta)
    @weights.each_with_index do |w, i|
      errors[i] += delta * w
    end
  end

  # Update the weights of the neuron during backpropagation
  def update_weights(inputs, delta, learning_rate)
    @weights = @weights.each_with_index.map do |weight, i|
      weight + learning_rate * delta * inputs[i]
    end

    @bias = @bias + learning_rate * delta
  end

end